### PR TITLE
New headless mode feature #105

### PR DESCRIPTION
**Pull request by reisfmb**
**Summary:**

My approach was to implement a lib "metadata" that abstracts the logic on the page processor.

That lib is then used on the page processor, without changing the logic previously defined.

This same lib is also used in a new "metadata" service, which purpose is to provide an endpoint that mimics the logic on the page processor having in mind the headless approach, i.e, returns a `json` in which the frontend can consume and inject the proper metadata:

image

Finally, to avoid hydration errors, a new "headless mode" checkbox config was set. If set to false (default value), then the app will work in the same way as it was. If set to true, then page processor will do nothing and the "metadata" service will be available:

image

### CODE REVIEW

**Comment by reisfmb** [comment]
Some points for improvement:
- Improve name of [this function](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R135). Suggestions?
- Both [getMetaData](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R10) and [getTitle](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R77) depends on `site`, `siteContent` and `siteConfig`, which are returned from the get function. Since both are executed on the same scope on the page processor, we can improve it making both `getMetaData` and `getTitle` functions accept `site`, `siteContent` and `siteConfig` as props, and only run the current `get` function once.

**Comment by reisfmb** [comment]
[Here](https://github.com/enonic/site-enonic-next/blob/12f5efc617c2bb9fecab3161355050ef8c6221df/src/scripts/SeoMetaFieldsScript.tsx) is an example of the usage of the app on site-enonic-next

**Comment by alansemenov** [comment]
> Improve name of [this function](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R135). Suggestions?

`getContentAndSite` or `getContentProps`

> Both [getMetaData](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R10) and [getTitle](https://github.com/enonic/app-metafields/pull/106/commits/3aa3b27cfce3343690bd7ca69fbf6837bbfe47ca#diff-bcd5f018db4814ab199de033d8d549472d53b4acbdfab118d8fccb6cddb1f2d1R77) depends on site, siteContent and siteConfig, which are returned from the get function. Since both are executed on the same scope on the page processor, we can improve it making both getMetaData and getTitle functions accept site, siteContent and siteConfig as props, and only run the current get function once.

Yes, we should do that.


I would also stay clear of one-character variable names, unless it's a simple counter.

```
const g = get(contentPath);
```

Should be something like

```
const contentAndSite = getContentAndSite(contentPath);
```

**Comment by reisfmb** [comment]
@alansemenov 

Implemented those improvements + helpful console warnings in case of service failing to retrieve data.

Tested the usage of the app on site market locally, with headless mode enabled and disabled and it seems to be working just fine!

If approved, I think we're good to go with a release.

**Review by alansemenov — APPROVED** [review]

File: src/main/resources/site/site.xml [hunk 1]
```diff
@@ -7,6 +7,11 @@
 		<field-set>
 			<label>General settings</label>
 			<items>
+				<input name="headless" type="CheckBox">
+					<label>Headless mode</label>
+					<help-text>If enabled, page processor will do nothing and its logic will be available in "metadata" service.</help-text>
+					<occurrences minimum="0" maximum="1"/>
+				</input>
```
Comment (poi33) [inline src/main/resources/site/site.xml:14]: The frontend should disable page processing. As we do in the nextxp proxy: https://github.com/enonic/lib-nextxp/commit/fc72d3750740bee3914c4b8b96edc36ff92a03f8

Comment (reisfmb) [reply src/main/resources/site/site.xml:14]: Why?

Comment (poi33) [reply src/main/resources/site/site.xml:14]: You usually don't want the backend to modify your frontend html. (Currently headless often only does this inside xp/content studio rendering and not live)
The extension of this app is made to access the data for the frontend.

I guess it boils down to if you want to disable this on every app or one global place for all headless to disable. I think the fontend should deal with it.

**Comment by sigdestad** [comment]
I did not read through all the details here, but I'm confident we should _not_ have checkbox called "headless mode". Basically, the preview can prevent injections from this app by setting the "post processing = false" in the response object.

**Pull request merged**